File: evolution.jl
using MKL
using ITensors
using LinearAlgebra
include("evolve.jl")
include("methods.jl")

ITensors.Strided.set_num_threads(1)
MKL.BLAS.set_num_threads(1)
ITensors.enable_threaded_blocksparse()

# system size
N = 10
# energy parameters, in units eV
it = .52
iU = 0.5 * it

# lattice spacing, in angstroms
ia = 4

# pulse parameters
iF0 = 10  # field strength in MV/cm
iomega0 = 32.9  # driving (angular) frequency, in THz
cycles = 10

# CONVERTING TO ATOMIC UNITS, w/ energy normalized to t
factor = 1 / (it * 0.036749323)
t = 1
U = iU / it

omega0 = iomega0 * factor * 0.0001519828442
a = ia * 1.889726125/factor
F0 = iF0 * 1.944689151e-4 * (factor^2)

# create the local hilbert space on N sites
sites = siteinds("Electron", N; conserve_qns=true)

# set up parameters struct
params = Parameters(N, sites, U, a, cycles, omega0, F0)

# get ground hamiltonian as an MPO
H_ground = get_ham(0, params)

# Prepare initial state MPS
state = [isodd(n) ? "Up" : "Dn" for n=1:N]
psi0_i = productMPS(params.space, state)

# Do 8 sweeps of DMRG , gradually increasing the maximum MPS
# bond dimension, at 12 sites, this gives precision to 7 sig figs
sweeps = Sweeps(8)
maxdim!(sweeps, 10, 20, 100, 200, 400, 600, 800)
cutoff!(sweeps, 1e-10)
# Run the DMRG algorithm
energy, psi0 = @time dmrg(H_ground, psi0_i, sweeps)

@show energy

# times for evolution
nsteps = 2000
ti = 0
tf = 2 * pi * cycles / omega0
tau = (tf - ti) / nsteps  # time step
cutoff = 1E-8

psi = psi0

currents = zeros(nsteps)
energies = zeros(nsteps)

#Time evolution
@time for step=0:nsteps-1
    curr_time = step * tau
    phi_t = phi_tl(curr_time, a, F0, omega0, cycles)
    phi_td2 = phi_tl(curr_time + tau / 2, a, F0, omega0, cycles)
    phi_td = phi_tl(curr_time + tau, a, F0, omega0, cycles)
    htd2 = get_itensor_ham(N, sites, phi_td2, U)
    k1 = -1.0im * tau * apply(get_itensor_ham(N, sites, phi_t, U), psi)
    k2 = -1.0im * tau * apply(htd2, psi) + -1.0im * tau / 2 * apply(htd2, k1)
    k3 = -1.0im * tau * apply(htd2, psi) + -1.0im * tau / 2 * apply(htd2, k2)
    k4 = -1.0im * tau * apply(get_itensor_ham(N, sites, phi_td, U), psi + k3)
    global psi += (1/6) * k1 + (1/3) * k2 + (1/3) * k3 + (1/6) * k1
    global psi = (1 / norm(psi)) * psi
    # global psi = apply(get_prop_gates(N, sites, tau, phi, U), psi; cutoff=cutoff, maxdim=800)
    # calculate energy by taking <psi|H|psi>
    local current = inner(psi, get_current(N, sites, phi_td, a), psi)
    currents[step + 1] = real(current)
    local energy = inner(psi, get_ham(N, sites, phi_td, U), psi)
    energies[step + 1] = real(energy)
end

io = open("./Data/RK4Testing/mps-current-U$U-nsites$N-nsteps$nsteps.txt", "w")
for step=1:nsteps
    write(io, "$(currents[step])\n")
end
close(io)

io = open("./Data/RK4Testing/mps-energy-U$U-nsites$N-nsteps$nsteps.txt", "w")
for step=1:nsteps
    write(io, "$(energies[step])\n")
end
close(io)

File: evolution.py

"""Tenpy packages"""
from tenpy.tools.params import Config
from tenpy.algorithms.dmrg import TwoSiteDMRGEngine as DMRG
from tenpy.networks.mps import MPS
from tebd import Engine as TEBD
from tenpy.algorithms.truncation import TruncationError
from tools import *

import numpy as np
from matplotlib import pyplot as plt
import time

# energy parameters, in units eV
it = .52
##########################
"""IMPORTANT PARAMETERS"""
##########################
phi_func = phi_tl
maxerr = 1e-12  # used for DMRG
maxdim = 2400 # maximum bond dimension, used for TEBD
pbc = False
N = 20
iU = 1 * it
pbc = False  # periodic boundary conditions
nsteps = 2000

"""We will hold these parameters constant"""
# lattice spacing, in angstroms
ia = 4
# pulse parameters
iF0 = 10  # field strength in MV/cm
iomega0 = 32.9  # driving (angular) frequency, in THz
cycles = 10


p = Parameters(N, iU, it, ia, cycles, iomega0, iF0, pbc)

# get the start time
start_time = time.time()
model = FHHamiltonian(0, p, phi_func)
current = FHCurrent(0, p, phi_func)
sites = model.lat.mps_sites()
state = ["up", "down"] * (N // 2)
psi0_i = MPS.from_product_state(sites, state)

# the max bond dimension
chi_list = {0:20, 1:40, 2:100, 4:200, 6:400, 8: 1200, 10:maxdim}
dmrg_dict = {"chi_list":chi_list, "max_E_err":maxerr, "max_sweeps":10, "mixer":True, "combine":False}
dmrg_params = Config(dmrg_dict, "DMRG-maxerr{}".format(maxerr))
dmrg = DMRG(psi0_i, model, dmrg_params)
E, psi0 = dmrg.run()

psi = psi0

ti = 0
tf = 2 * np.pi * cycles / p.field
times, delta = np.linspace(ti, tf, num=nsteps, endpoint=True, retstep=True)
# we pass in nsteps - 1 because we would like to evauluate the system at
# nsteps time points, including the ground state calculations
tebd_dict = {"dt":delta, "order":2, "start_time":ti, "start_trunc_err":TruncationError(eps=maxerr), "trunc_params":{"svd_min":maxerr, "chi_max":maxdim}, "N_steps":nsteps-1, "verbose":0}
tebd_params = Config(tebd_dict, "TEBD-trunc_err{}-nsteps{}".format(maxerr, nsteps))
tebd = TEBD(psi, model, p, phi_tl, tebd_params)
times, energies, currents = tebd.run()

tot_time = time.time() - start_time

print("Evolution complete, total time:", tot_time)

savedir = "./Data/Tenpy/Basic/"
allps = "-nsteps{}".format(nsteps)
ecps = "-nsites{}-U{}-maxdim{}-maxerr{}".format(p.nsites, p.u, maxdim, maxerr)
np.save(savedir + "times" + allps + ".npy", times)
np.save(savedir + "energies" + allps + ecps + ".npy", energies)
np.save(savedir + "currents" + allps + ecps + ".npy", currents)

# write metadata to file (evolution time and error)
with open(savedir + "metadata" + allps + ecps + ".txt", "w") as f:
    f.write(str(tot_time) + "\n")
